api/handlers.go

```diff
@@ -0,0 +1,156 @@
+package server
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"strconv"
+	"strings"
+)
+
+// GoCertRouter takes in an environment struct, passes it along to any handlers that will need
+// access to it, then builds and returns it for a server to consume
+func GoCertRouter(env *Environment) http.Handler {
+	router := http.NewServeMux()
+	router.HandleFunc("GET /certificate_requests", GetCertificateRequests(env))
+	router.HandleFunc("POST /certificate_requests", PostCertificateRequest(env))
+	router.HandleFunc("GET /certificate_requests/{id}", GetCertificateRequest(env))
+	router.HandleFunc("DELETE /certificate_requests/{id}", DeleteCertificateRequest(env))
+	router.HandleFunc("POST /certificate_requests/{id}/certificate", PostCertificate(env))
+
+	v1 := http.NewServeMux()
+	v1.HandleFunc("GET /status", HealthCheck)
+	v1.Handle("/api/v1/", http.StripPrefix("/api/v1", router))
+
+	return logging(v1)
+}
+
+// the health check endpoint simply returns a 200
+func HealthCheck(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(200) //nolint:errcheck
+}
+
+// GetCertificateRequests returns all of the Certificate Requests
+func GetCertificateRequests(env *Environment) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		certs, err := env.DB.RetrieveAll()
+		if err != nil {
+			logError(err.Error(), 500, w)
+			return
+		}
+		body, err := json.Marshal(certs)
+		if err != nil {
+			logError(err.Error(), 500, w)
+			return
+		}
+		if _, err := w.Write(body); err != nil {
+			logError(err.Error(), 500, w)
+		}
+	}
+}
+
+// PostCertificateRequest creates a new Certificate Request, and returns the id of the created row
+func PostCertificateRequest(env *Environment) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		csr, err := io.ReadAll(r.Body)
+		if err != nil {
+			logError(err.Error(), 500, w)
+			return
+		}
+		id, err := env.DB.Create(string(csr))
+		if err != nil {
+			if strings.Contains(err.Error(), "UNIQUE constraint failed") {
+				logError("given csr already recorded", 400, w)
+				return
+			} else {
+				logError(err.Error(), 400, w)
```

Comment:
I think this might be good enough for now, but I assume some errors would be due to the content of the request, and some would be due to something on our side, like the DB not responding for some reason. We might have to revisit this to return a 5xx if the error is on our side. Additionally, if the error is with the content sent to us, we could return 415 or 422.

This could be out of scope for this PR however and evaluated as part of another task.

Reply:
This line is a real error, thanks for catching it.

I've done my best to differentiate the two cases, but excluding the errors I couldn't forsee, I think these responses are "production" "ready". In my experience it's generally a bad idea to use rarely used 400's because the consumers of API's like to switch over error messages. When it comes to errors about the HTTP request directly, Go will send the proper expected responses in those situations before it ever gets to the handler code.

Like you said though, we can always come back to this later.

Maybe we can send a 418 for the health check :D

Reply:
Please no!